File: server/Server/src/me/spencernold/server/util/ByteBuffer.java
package me.spencernold.server.util;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Optional;

public class ByteBuffer {

	private final ByteArrayOutputStream output;
	private final ByteArrayInputStream input;
	
	public ByteBuffer(byte[] bytes) {
		input = new ByteArrayInputStream(bytes);
		output = null;
	}
	
	public ByteBuffer() {
		output = new ByteArrayOutputStream();
		input = null;
	}
	
	public void writeBytes(byte[] bytes) {
		try {
			output.write(bytes);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	public void writeNBytes(byte[] bytes) {
		writeInt(bytes.length);
		writeBytes(bytes);
	}
	
	public void writeByte(byte b) {
		writeBytes(new byte[] {b});
	}
	
	public void writeBoolean(boolean b) {
		writeByte((byte) (b ? 1 : 0));
	}
	
	public void writeShort(short s) {
		writeBytes(Serializer.shortToBytesBE(s));
	}
	
	public void writeInt(int i) {
		writeBytes(Serializer.intToBytesBE(i));
	}
	
	public void writeLong(long l) {
		writeBytes(Serializer.longToBytesBE(l));
	}
	
	public void writeFloat(float f) {
		writeInt(Float.floatToIntBits(f));
	}
	
	public void writeDouble(double d) {
		writeLong(Double.doubleToLongBits(d));
	}
	
	public void writeString(String s) {
		writeInt(s.length());
		writeBytes(s.getBytes(StandardCharsets.UTF_8));
	}
	
	public void writeOptionalString(Optional<String> optional) {
		boolean b = optional.isPresent();
		writeBoolean(b);
		if (b)
			writeString(optional.get());
	}
	
	public byte[] readBytes(int n) {
		try {
			return input.readNBytes(n);
		} catch (IOException e) {
			e.printStackTrace();
			return new byte[0];
		}
	}
	
	public byte[] readNBytes() {
		int length = readInt();
		return readBytes(length);
	}
	
	public byte readByte() {
		byte[] bytes = readBytes(1);
		if (bytes.length != 1)
			return (byte) 0;
		return bytes[0];
	}
	
	public boolean readBoolean() {
		return readByte() == 1;
	}
	
	public short readShort() {
		byte[] bytes = readBytes(2);
		if (bytes.length != 2)
			return (short) 0;
		return Serializer.bytesToShortBE(bytes);
	}
	
	public int readInt() {
		byte[] bytes = readBytes(4);
		if (bytes.length != 4)
			return 0;
		return Serializer.bytesToIntBE(bytes);
	}
	
	public long readLong() {
		byte[] bytes = readBytes(8);
		if (bytes.length != 8)
			return 0;
		return Serializer.bytesToLongBE(bytes);
	}
	
	public float readFloat() {
		return Float.intBitsToFloat(readInt());
	}
	
	public double readDouble() {
		return Double.longBitsToDouble(readLong());
	}
	
	public String readString() {
		int n = readInt();
		return new String(readBytes(n), StandardCharsets.UTF_8);
	}
	
	public Optional<String> readOptionalString() {
		boolean b = readBoolean();
		if (b)
			return Optional.of(readString());
		return Optional.empty();
	}
	
	public byte[] toByteArray() {
		return output == null ? input.readAllBytes() : output.toByteArray();
	}
}

File: client/lib/utils/bytebuffer.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:client/utils/serializer.dart';

class ByteBuf {
  List<int> _data;

  ByteBuf.read(this._data);

  ByteBuf.write() : _data = [];

  void writeBytes(List<int> bytes) {
    _data.addAll(bytes);
  }

  void writeByte(int i) {
    writeBytes([i]);
  }

  void writeBoolean(bool b) {
    writeByte(b ? 1 : 0);
  }

  void writeShort(int i) {
    writeBytes(Serializer.shortToBytesBE(i));
  }

  void writeInt(int i) {
    writeBytes(Serializer.intToBytesBE(i));
  }

  void writeLong(int i) {
    writeBytes(Serializer.longToBytesBE(i));
  }

  void writeFloat(double d) {
    writeInt(Serializer.floatBitsToInt(d));
  }

  void writeDouble(double d) {
    writeLong(Serializer.doubleBitsToLong(d));
  }

  void writeString(String s) {
    writeInt(s.length);
    writeBytes(utf8.encode(s));
  }

  Uint8List readBytes(int n) {
    List<int> bytes = _data.sublist(0, n);
    _data = _data.sublist(n);
    return Uint8List.fromList(bytes);
  }

  int readByte() {
    return readBytes(1)[0];
  }

  bool readBoolean() {
    return readByte() == 1;
  }

  int readShort() {
    return Serializer.bytesToShortBE(readBytes(2));
  }

  int readInt() {
    return Serializer.bytesToIntBE(readBytes(4));
  }

  int readLong() {
    return Serializer.bytesToLongBE(readBytes(8));
  }

  double readFloat() {
    return Serializer.intBitsToFloat(readInt());
  }

  double readDouble() {
    return Serializer.longBitsToDouble(readLong());
  }

  String readString() {
    int size = readInt();
    return utf8.decode(readBytes(size));
  }

  Uint8List toByteArray() {
    return Uint8List.fromList(_data);
  }

  bool hasMoreData() {
    return _data.isNotEmpty;
  }

  int bytes() {
    return _data.length;
  }
}
